File: foreign/csharp/Iggy_SDK/Vsr/LoginRegister.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using System.Buffers.Binary;
using System.Text;
using Apache.Iggy.Utils;

namespace Apache.Iggy.Vsr;

/// <summary>
///     The register handshake bodies. VSR replaces the legacy login codes with
///     <c>LOGIN_REGISTER</c> / <c>LOGIN_REGISTER_WITH_PAT</c>, whose bodies lead with the client version
///     info the server gates on before it touches credentials.
/// </summary>
internal static class LoginRegister
{
    internal const string SDK_NAME = "csharp-sdk";

    /// <summary>Semver of the <c>iggy_binary_protocol</c> crate this SDK is built against.</summary>
    internal const int PROTOCOL_VERSION_MAJOR = 0;

    internal const int PROTOCOL_VERSION_MINOR = 11;
    internal const int PROTOCOL_VERSION_PATCH = 0;

    /// <summary>Packed protocol version: <c>major &lt;&lt; 20 | minor &lt;&lt; 10 | patch</c>, 10 bits each.</summary>
    internal const uint PROTOCOL_VERSION =
        ((uint)PROTOCOL_VERSION_MAJOR << 20) | ((uint)PROTOCOL_VERSION_MINOR << 10) | PROTOCOL_VERSION_PATCH;

    private const int MaxWireNameLength = 255;

    private static int VersionInfoLength => 4 + NameLength(SDK_NAME) + NameLength(SdkVersion.Value);

    internal static byte[] Serialize(string username, string password, string? clientContext = null)
    {
        CredentialBounds.ValidateUsername(username);
        CredentialBounds.ValidatePassword(password);

        var writer = new BodyWriter(VersionInfoLength + NameLength(username) + NameLength(password) + 4 +
                                    ContextLength(clientContext));
        writer.WriteVersionInfo();
        writer.WriteName(username, nameof(username));
        writer.WriteName(password, nameof(password));
        writer.WriteContext(clientContext);

        return writer.Buffer;
    }

    internal static byte[] SerializeWithPersonalAccessToken(string token, string? clientContext = null)
    {
        CredentialBounds.ValidateToken(token);

        var writer = new BodyWriter(VersionInfoLength + NameLength(token) + 4 + ContextLength(clientContext));
        writer.WriteVersionInfo();
        writer.WriteName(token, nameof(token));
        writer.WriteContext(clientContext);

        return writer.Buffer;
    }

    /// <summary>
    ///     <c>[user_id u32][session u64][server_protocol_version u32][server_version len u8 + bytes]</c>.
    /// </summary>
    internal static LoginRegisterResponse Deserialize(ReadOnlySpan<byte> body)
    {
        if (body.IsEmpty)
        {
            // The server fast-fails a terminal register failure (invalid credentials, invalid token,
            // inactive user) with an empty reply instead of a typed error frame, and the reason is not
            // recoverable from the wire. Same INVALID_FORMAT surface as the Rust SDK.
            throw VsrError.Exception(VsrError.INVALID_FORMAT,
                "Server rejected the login. The register reply is empty, which the server sends for invalid " +
                "credentials, an invalid personal access token, or an inactive user.");
        }

        if (body.Length < 17)
        {
            throw VsrError.Exception(VsrError.INVALID_FORMAT, "Register reply is truncated.");
        }

        var serverVersionLength = body[16];
        if (serverVersionLength == 0 || body.Length < 17 + serverVersionLength)
        {
            throw VsrError.Exception(VsrError.INVALID_FORMAT, "Register reply carries a malformed server version.");
        }

        return new LoginRegisterResponse(BinaryPrimitives.ReadUInt32LittleEndian(body[..4]),
            BinaryPrimitives.ReadUInt64LittleEndian(body[4..12]),
            BinaryPrimitives.ReadUInt32LittleEndian(body[12..16]),
            Encoding.UTF8.GetString(body.Slice(17, serverVersionLength)));
    }

    private static int NameLength(string value)
    {
        return 1 + Encoding.UTF8.GetByteCount(value);
    }

    private static int ContextLength(string? clientContext)
    {
        return clientContext is null ? 0 : Encoding.UTF8.GetByteCount(clientContext);
    }

    private struct BodyWriter
    {
        private int _position;

        internal BodyWriter(int length)
        {
            Buffer = new byte[length];
            _position = 0;
        }

        internal byte[] Buffer { get; }

        internal void WriteVersionInfo()
        {
            BinaryPrimitives.WriteUInt32LittleEndian(Buffer.AsSpan(_position, 4), PROTOCOL_VERSION);
            _position += 4;
            WriteName(SDK_NAME, nameof(SDK_NAME));
            WriteName(SdkVersion.Value, nameof(SdkVersion));
        }

        internal void WriteName(string value, string name)
        {
            var length = Encoding.UTF8.GetByteCount(value);
            if (length is 0 or > MaxWireNameLength)
            {
                throw new ArgumentException($"{name} must be 1-{MaxWireNameLength} UTF-8 bytes, got {length}.", name);
            }

            Buffer[_position] = (byte)length;
            _position += 1;
            Encoding.UTF8.GetBytes(value, Buffer.AsSpan(_position, length));
            _position += length;
        }

        internal void WriteContext(string? clientContext)
        {
            var length = ContextLength(clientContext);
            BinaryPrimitives.WriteUInt32LittleEndian(Buffer.AsSpan(_position, 4), (uint)length);
            _position += 4;
            if (clientContext is not null && length > 0)
            {
                Encoding.UTF8.GetBytes(clientContext, Buffer.AsSpan(_position, length));
                _position += length;
            }
        }
    }
}

internal readonly record struct LoginRegisterResponse(
    uint UserId,
    ulong Session,
    uint ServerProtocolVersion,
    string ServerVersion);

File: core/binary_protocol/src/responses/users/login_register.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use crate::WireError;
use crate::codec::{WireDecode, WireEncode, read_u32_le, read_u64_le};
use crate::primitives::identifier::WireName;
use bytes::{BufMut, BytesMut};

/// Combined login + register response for the server.
///
/// Returns the authenticated user's ID, the consensus session number
/// (commit op number from the Register operation), and the server's
/// protocol version + build version so both sides know what they talk to.
///
/// Wire format:
/// ```text
/// [user_id:u32 LE][session:u64 LE][server_protocol_version:u32 LE]
/// [server_version_len:u8][server_version:N]
/// ```
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LoginRegisterResponse {
    pub user_id: u32,
    pub session: u64,
    pub server_protocol_version: u32,
    pub server_version: WireName,
}

impl WireEncode for LoginRegisterResponse {
    fn encoded_size(&self) -> usize {
        16 + self.server_version.encoded_size()
    }

    fn encode(&self, buf: &mut BytesMut) {
        buf.put_u32_le(self.user_id);
        buf.put_u64_le(self.session);
        buf.put_u32_le(self.server_protocol_version);
        self.server_version.encode(buf);
    }
}

impl WireDecode for LoginRegisterResponse {
    fn decode(buf: &[u8]) -> Result<(Self, usize), WireError> {
        let user_id = read_u32_le(buf, 0)?;
        let session = read_u64_le(buf, 4)?;
        let server_protocol_version = read_u32_le(buf, 12)?;
        let (server_version, server_version_len) = WireName::decode(&buf[16..])?;
        Ok((
            Self {
                user_id,
                session,
                server_protocol_version,
                server_version,
            },
            16 + server_version_len,
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn sample() -> LoginRegisterResponse {
        LoginRegisterResponse {
            user_id: 42,
            session: 100,
            server_protocol_version: 1,
            server_version: WireName::new("0.8.0").unwrap(),
        }
    }

    #[test]
    fn roundtrip() {
        let resp = sample();
        let bytes = resp.to_bytes();
        let (decoded, consumed) = LoginRegisterResponse::decode(&bytes).unwrap();
        assert_eq!(consumed, bytes.len());
        assert_eq!(decoded, resp);
    }

    #[test]
    fn encoded_size_matches_output() {
        let resp = sample();
        assert_eq!(resp.encoded_size(), resp.to_bytes().len());
    }

    #[test]
    fn truncated_returns_error() {
        let bytes = sample().to_bytes();
        for i in 0..bytes.len() {
            assert!(
                LoginRegisterResponse::decode(&bytes[..i]).is_err(),
                "expected error for truncation at byte {i}"
            );
        }
    }

    #[test]
    fn wire_layout() {
        let resp = LoginRegisterResponse {
            user_id: 0x0102_0304,
            session: 0x0506_0708_090A_0B0C,
            server_protocol_version: 0x0D0E_0F10,
            server_version: WireName::new("v").unwrap(),
        };
        let bytes = resp.to_bytes();
        assert_eq!(
            u32::from_le_bytes(bytes[..4].try_into().unwrap()),
            0x0102_0304
        );
        assert_eq!(
            u64::from_le_bytes(bytes[4..12].try_into().unwrap()),
            0x0506_0708_090A_0B0C
        );
        assert_eq!(
            u32::from_le_bytes(bytes[12..16].try_into().unwrap()),
            0x0D0E_0F10
        );
        assert_eq!(bytes[16], 1);
        assert_eq!(bytes[17], b'v');
    }
}
